File: ctx_trend/ctx_trend.py
"""C-Telopeptide (CTX) trend by PIN and visit.

Reads a Labcorp study-results export in the "Study Data Entry Template" layout
and pulls the serum C-Telopeptide column into a tidy long table keyed by
participant (PIN) and visit, then writes a wide PIN x Visit table and a trend
plot.

Layout assumed (matches Study_Data_Entry_Template_..._Labcorp_Results.csv):
    row 1  column headers
    row 2  units             (C-Telopeptide,Serum -> pg/mL)
    row 3  reference intervals
    row 4+ one row per participant / visit / draw

Usage:
    python ctx_trend.py --csv path/to/Labcorp_Results.csv --outdir out
    python ctx_trend.py --csv path/to/Labcorp_Results.csv --x days
"""

from __future__ import annotations

import argparse
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import pandas as pd

CTX_COL = "C-Telopeptide,Serum"
UNITS_ROW = 0        # 0-based position within the two non-data header rows
REF_ROW = 1
ID_COLS = ["PIN", "Visit_Number", "Pre_Post_Old", "Pre_Post_New", "Collection_Date"]


def load_labs(csv_path: str | Path) -> tuple[pd.DataFrame, dict[str, str], dict[str, str]]:
    """Return (data, units, reference_intervals) from the template CSV.

    The two metadata rows under the header are stripped off and returned as
    {column -> value} dicts so the CTX units/ref range can be used in labels.
    """
    raw = pd.read_csv(csv_path, dtype=str, keep_default_na=False)
    units = raw.iloc[UNITS_ROW].to_dict()
    refs = raw.iloc[REF_ROW].to_dict()
    data = raw.iloc[REF_ROW + 1 :].copy()

    # Drop the blank filler rows at the bottom of the template.
    data = data[data["PIN"].str.strip() != ""]
    return data.reset_index(drop=True), units, refs


def ctx_by_pin_visit(data: pd.DataFrame) -> pd.DataFrame:
    """Tidy one-row-per-draw CTX table with baseline-relative columns.

    Participants 001-006 record the pre/post draw flag in ``Pre_Post_Old`` and
    007+ in ``Pre_Post_New``; the two are coalesced into a single ``pre_post``.
    A visit can appear twice for the same PIN (a pre and a post draw), so the
    natural key is (pin, visit, pre_post).
    """
    ctx = pd.DataFrame(
        {
            "pin": data["PIN"].str.strip(),
            "visit": pd.to_numeric(data["Visit_Number"], errors="coerce").astype("Int64"),
            "pre_post": pd.to_numeric(
                data["Pre_Post_New"].where(data["Pre_Post_New"].str.strip() != "", data["Pre_Post_Old"]),
                errors="coerce",
            ).astype("Int64"),
            "collection_date": pd.to_datetime(data["Collection_Date"], errors="coerce"),
            "ctx_pg_ml": pd.to_numeric(data[CTX_COL], errors="coerce"),
        }
    )

    # Keep only draws that actually have a CTX result.
    ctx = ctx.dropna(subset=["ctx_pg_ml"]).sort_values(["pin", "visit", "pre_post"])

    ctx["draw"] = ctx["pre_post"].map({0: "pre", 1: "post"}).fillna("single")
    ctx["visit_label"] = [
        f"V{v}" if d == "single" else f"V{v} ({d})" for v, d in zip(ctx["visit"], ctx["draw"])
    ]

    # Baseline = each participant's earliest CTX draw.
    baseline_value = ctx.groupby("pin")["ctx_pg_ml"].transform("first")
    baseline_date = ctx.groupby("pin")["collection_date"].transform("first")
    ctx["baseline_pg_ml"] = baseline_value
    ctx["change_from_baseline"] = ctx["ctx_pg_ml"] - baseline_value
    ctx["pct_change_from_baseline"] = 100 * ctx["change_from_baseline"] / baseline_value
    ctx["days_from_baseline"] = (ctx["collection_date"] - baseline_date).dt.days

    return ctx.reset_index(drop=True)


def ctx_wide(ctx: pd.DataFrame) -> pd.DataFrame:
    """PIN x visit matrix of CTX values (rows = PIN, columns = visit label)."""
    wide = ctx.pivot_table(
        index="pin", columns="visit_label", values="ctx_pg_ml", aggfunc="first"
    )
    # Order columns by visit number, then pre before post.
    order = (
        ctx[["visit_label", "visit", "pre_post"]]
        .drop_duplicates()
        .sort_values(["visit", "pre_post"])["visit_label"]
    )
    return wide.reindex(columns=[c for c in order if c in wide.columns])


def plot_ctx_trend(
    ctx: pd.DataFrame,
    out_path: str | Path,
    x: str = "visit",
    units: str = "pg/mL",
    ref_range: str = "",
) -> Path:
    """Line plot of CTX over time, one series per PIN."""
    x_col = {"visit": "visit", "days": "days_from_baseline"}[x]
    x_label = {"visit": "Visit number", "days": "Days from baseline draw"}[x]

    fig, ax = plt.subplots(figsize=(9, 5.5))
    for pin, grp in ctx.groupby("pin"):
        grp = grp.sort_values(x_col)
        ax.plot(grp[x_col], grp["ctx_pg_ml"], marker="o", linewidth=1.8, label=f"PIN {pin}")

    if ref_range and "-" in ref_range:
        low, high = (float(v) for v in ref_range.split("-", 1))
        ax.axhspan(low, high, color="0.85", zorder=0, label=f"Reference {ref_range} {units}")

    ax.set_xlabel(x_label)
    ax.set_ylabel(f"C-Telopeptide, serum ({units})")
    ax.set_title("C-Telopeptide trend by participant and visit")
    if x == "visit":
        ax.set_xticks(sorted(ctx["visit"].dropna().unique()))
    ax.grid(axis="y", alpha=0.3)
    ax.legend(title="Participant", frameon=False)
    fig.tight_layout()

    out_path = Path(out_path)
    fig.savefig(out_path, dpi=150)
    plt.close(fig)
    return out_path


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--csv", required=True, help="Labcorp results CSV in the study template layout")
    parser.add_argument("--outdir", default="out", help="directory for the CSV/PNG outputs")
    parser.add_argument("--x", choices=["visit", "days"], default="visit", help="x-axis for the trend plot")
    args = parser.parse_args()

    outdir = Path(args.outdir)
    outdir.mkdir(parents=True, exist_ok=True)

    data, units, refs = load_labs(args.csv)
    ctx = ctx_by_pin_visit(data)
    wide = ctx_wide(ctx)

    long_path = outdir / "ctx_by_pin_visit.csv"
    wide_path = outdir / "ctx_pin_by_visit_wide.csv"
    ctx.to_csv(long_path, index=False)
    wide.to_csv(wide_path)

    plot_path = plot_ctx_trend(
        ctx,
        outdir / "ctx_trend.png",
        x=args.x,
        units=units.get(CTX_COL, "pg/mL"),
        ref_range=refs.get(CTX_COL, ""),
    )

    print(f"CTX results: {len(ctx)} draws across {ctx['pin'].nunique()} participants\n")
    print(
        ctx[["pin", "visit", "draw", "collection_date", "ctx_pg_ml", "pct_change_from_baseline"]].to_string(
            index=False
        )
    )
    print("\nPIN x visit matrix (pg/mL):")
    print(wide.to_string())
    print(f"\nWrote {long_path}\n      {wide_path}\n      {plot_path}")


if __name__ == "__main__":
    main()

File: ctx_trend/ctx_trend.R
#!/usr/bin/env Rscript
#
# C-Telopeptide (CTX) trend by PIN and visit.
#
# Works on a data frame already loaded in your session — named `study` below —
# holding a Labcorp study-results export in the "Study Data Entry Template"
# layout. Pulls the serum C-Telopeptide column into a tidy long table keyed by
# participant (PIN) and visit, then builds a wide PIN x Visit table and a trend
# plot.
#
# Layout assumed (matches Study_Data_Entry_Template_..._Labcorp_Results.csv):
#     row 1  units             (C-Telopeptide,Serum -> pg/mL)
#     row 2  reference intervals
#     row 3+ one row per participant / visit / draw
#
# Those first two rows sit under the header, so they arrive as data rows in
# `study` no matter how it was read; `ctx_by_pin_visit()` drops them, along with
# the blank filler rows at the bottom of the template.
#
# Usage:
#     source("ctx_trend.R")
#     ctx  <- ctx_by_pin_visit(study)
#     wide <- ctx_wide(ctx)
#     plot_ctx_trend(ctx)                 # x = visit number
#     plot_ctx_trend(ctx, x = "days")     # x = days from baseline draw

suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(ggplot2)
})

#' Locate a column by name, tolerating however `study` was read in.
#'
#' `read.csv()` turns "C-Telopeptide,Serum" into "C.Telopeptide.Serum" and
#' `janitor::clean_names()` into "c_telopeptide_serum", while `readr::read_csv()`
#' keeps it verbatim. Comparing on alphanumerics only makes all of these match.
find_col <- function(df, name) {
  norm <- function(x) tolower(gsub("[^A-Za-z0-9]", "", x))
  hit <- which(norm(names(df)) == norm(name))
  if (length(hit) == 0) {
    stop(sprintf("column '%s' not found in the data frame", name), call. = FALSE)
  }
  names(df)[hit[1]]
}

#' Pull a column as trimmed character, blanks as NA.
#'
#' Columns may already be typed (numeric/Date) if the reader guessed, so
#' everything is routed through character first for one coercion path.
col_chr <- function(df, name) {
  x <- trimws(as.character(df[[find_col(df, name)]]))
  dplyr::na_if(x, "")
}

#' Tidy one-row-per-draw CTX table with baseline-relative columns.
#'
#' Participants 001-006 record the pre/post draw flag in `Pre_Post_Old` and
#' 007+ in `Pre_Post_New`; the two are coalesced into a single `pre_post`.
#' A visit can appear twice for the same PIN (a pre and a post draw), so the
#' natural key is (pin, visit, pre_post).
#'
#' @param study data frame of the export, header row already used as names
ctx_by_pin_visit <- function(study) {
  # Drop the units / reference-interval rows and the blank filler rows first, so
  # the header text ("yyyy-mm-dd", "pg/mL") never reaches the type coercions.
  pin <- col_chr(study, "PIN")
  study <- study[!is.na(pin) & !grepl("^(ParticipantID|ReferenceIntervals)", pin), ]

  ctx <- tibble(
    pin = col_chr(study, "PIN"),
    visit = suppressWarnings(as.integer(col_chr(study, "Visit_Number"))),
    pre_post = suppressWarnings(as.integer(coalesce(
      col_chr(study, "Pre_Post_New"),
      col_chr(study, "Pre_Post_Old")
    ))),
    collection_date = as.Date(col_chr(study, "Collection_Date"), format = "%Y-%m-%d"),
    ctx_pg_ml = suppressWarnings(as.numeric(col_chr(study, "C-Telopeptide,Serum")))
  )

  ctx %>%
    # Keep only draws that actually have a CTX result.
    filter(!is.na(ctx_pg_ml)) %>%
    arrange(pin, visit, pre_post) %>%
    mutate(
      draw = case_when(
        pre_post == 0 ~ "pre",
        pre_post == 1 ~ "post",
        TRUE ~ "single"
      ),
      visit_label = ifelse(draw == "single",
                           paste0("V", visit),
                           paste0("V", visit, " (", draw, ")"))
    ) %>%
    # Baseline = each participant's earliest CTX draw.
    group_by(pin) %>%
    mutate(
      baseline_pg_ml = first(ctx_pg_ml),
      change_from_baseline = ctx_pg_ml - baseline_pg_ml,
      pct_change_from_baseline = 100 * change_from_baseline / baseline_pg_ml,
      days_from_baseline = as.integer(collection_date - first(collection_date))
    ) %>%
    ungroup()
}

#' Visit labels ordered by visit number, then pre before post.
visit_label_levels <- function(ctx) {
  ctx %>%
    distinct(visit_label, visit, pre_post) %>%
    arrange(visit, pre_post) %>%
    pull(visit_label)
}

#' PIN x visit matrix of CTX values (rows = PIN, columns = visit label).
ctx_wide <- function(ctx) {
  ctx %>%
    select(pin, visit_label, ctx_pg_ml) %>%
    pivot_wider(names_from = visit_label, values_from = ctx_pg_ml,
                values_fn = first) %>%
    arrange(pin) %>%
    select(pin, any_of(visit_label_levels(ctx)))
}

#' Units / reference interval for a lab column, read off the template's row 1-2.
#'
#' Returns "" when the export leaves the cell blank, as it does for CTX.
lab_metadata <- function(study, name = "C-Telopeptide,Serum") {
  pin <- col_chr(study, "PIN")
  value_at <- function(prefix) {
    row <- which(grepl(prefix, pin))[1]
    if (is.na(row)) return("")
    val <- as.character(study[[find_col(study, name)]][row])
    if (is.na(val)) "" else trimws(val)
  }
  list(units = value_at("^ParticipantID"), ref_range = value_at("^ReferenceIntervals"))
}

#' Line plot of CTX over time, one series per PIN.
plot_ctx_trend <- function(ctx, x = c("visit", "days"), units = "pg/mL",
                           ref_range = "") {
  x <- match.arg(x)
  x_col <- switch(x, visit = "visit", days = "days_from_baseline")
  x_label <- switch(x, visit = "Visit number", days = "Days from baseline draw")

  p <- ggplot(ctx, aes(x = .data[[x_col]], y = ctx_pg_ml,
                       colour = pin, group = pin))

  # Shade the reference interval when the export supplies one for CTX.
  if (nzchar(ref_range) && grepl("-", ref_range, fixed = TRUE)) {
    bounds <- as.numeric(strsplit(ref_range, "-", fixed = TRUE)[[1]])
    p <- p + annotate("rect", xmin = -Inf, xmax = Inf,
                      ymin = bounds[1], ymax = bounds[2],
                      fill = "grey85", alpha = 0.6)
  }

  p <- p +
    geom_line(linewidth = 0.8) +
    geom_point(size = 2.4) +
    labs(
      title = "C-Telopeptide trend by participant and visit",
      x = x_label,
      y = paste0("C-Telopeptide, serum (", units, ")"),
      colour = "Participant"
    ) +
    theme_minimal(base_size = 12) +
    theme(panel.grid.minor = element_blank())

  if (x == "visit") {
    p <- p + scale_x_continuous(breaks = sort(unique(ctx$visit)))
  }

  p
}

# ---------------------------------------------------------------------------
# Run on `study`
# ---------------------------------------------------------------------------
# Guarded so `source("ctx_trend.R")` just defines the functions when `study`
# is not (yet) in the session.

if (exists("study")) {
  ctx <- ctx_by_pin_visit(study)
  wide <- ctx_wide(ctx)
  meta <- lab_metadata(study)

  ctx_plot <- plot_ctx_trend(
    ctx,
    x = "visit",
    units = if (nzchar(meta$units)) meta$units else "pg/mL",
    ref_range = meta$ref_range
  )

  cat(sprintf("CTX results: %d draws across %d participants\n\n",
              nrow(ctx), n_distinct(ctx$pin)))
  print(as.data.frame(ctx[, c("pin", "visit", "draw", "collection_date",
                              "ctx_pg_ml", "pct_change_from_baseline")]),
        row.names = FALSE)
  cat("\nPIN x visit matrix (pg/mL):\n")
  print(as.data.frame(wide), row.names = FALSE)

  print(ctx_plot)

  # To save:
  # write.csv(ctx, "ctx_by_pin_visit.csv", row.names = FALSE, na = "")
  # write.csv(wide, "ctx_pin_by_visit_wide.csv", row.names = FALSE, na = "")
  # ggsave("ctx_trend.png", ctx_plot, width = 9, height = 5.5, dpi = 150)
}
